### ISSUE TITLE
System.Collections.Generic.KeyNotFoundException : The given key 'Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope' was not present in the dictionary.

### ISSUE THREAD

**Issue by sanjeev-chevireddy** [reqnroll/Reqnroll#175]
### Reqnroll Version

2.0.2

### Which test runner are you using?

NUnit

### Test Runner Version Number

4.0.1

### .NET Implementation

.NET 6.0

### Test Execution Method

Visual Studio Test Explorer

### Issue Description

 System.Collections.Generic.KeyNotFoundException : The given key 'Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope' was not present in the dictionary.
TearDown : System.Collections.Generic.KeyNotFoundException : The given key 'Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope' was not present in the dictionary.

  Stack Trace: 
ConcurrentDictionary`2.ThrowKeyNotFoundException(TKey key)
ConcurrentDictionary`2.get_Item(TKey key)
<>c.<RegisterProxyBindings>b__9_21(IServiceProvider sp)
CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
CallSiteRuntimeResolver.VisitDisposeCache(ServiceCallSite transientCallSite, RuntimeResolverContext context)
CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
<59 more frames...>
TestExecutionEngine.FireEventsAsync(HookType hookType)
TestExecutionEngine.FireScenarioEventsAsync(HookType bindingEvent)
TestExecutionEngine.OnScenarioEndAsync()
TestRunner.OnScenarioEndAsync()
CreateClaimFeature.TestTearDownAsync()
GenericAdapter`1.GetResult()
AsyncToSyncAdapter.Await[TResult](Func`1 invoke)
AsyncToSyncAdapter.Await(Func`1 invoke)
SetUpTearDownItem.RunSetUpOrTearDownMethod(TestExecutionContext context, IMethodInfo method)
SetUpTearDownItem.RunTearDown(TestExecutionContext context)

### Steps to Reproduce

We are currently migrating from SpecFlow to ReQnRoll and have encountered the issue when running the tests after migrating from SolidToken.SpecFlow.DependencyInjection to Reqnroll.Microsoft.Extensions.DependencyInjection;

Please see our Startup method as shown below

using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Reqnroll.Microsoft.Extensions.DependencyInjection;

 [ScenarioDependencies]
 public static IServiceCollection ConfigureServices()
 {
var configuration = new ConfigurationBuilder().AddJsonFile("appsettings.json").Build();
var connectionString = configuration["ConnectionString"];

IServiceCollection services = new ServiceCollection();
return services
    .AddSingleton(ConfigReader.GetBrowserSettingsData())
    .AddSingleton(ConfigReader.GetAppSettingsData<AppSettings>())
    .AddSingleton<IReportContext, ReportContext>()
    .AddSingleton<INotificationHelper, NotificationHelper>()
    .AddSingleton<IReportHelper, ReportHelper>()
    .AddSingleton<IPlaywrightDriver, PlaywrightDriver>()
    .AddSingleton<IHelper, Helper>()
    .AddSingleton<IApiContext, ApiContext>()
    .AddSingleton<IDataContext, DataContext>()
    .AddSingleton<ILoginPage, LoginPage>()
    .AddSingleton<IDbContainerFactory, CosmosContainerFactory>()
    .AddSingleton<IDbContainerFactory>(provider =>
    {
        //Authorisation
        var client = GetCosmosClient(connectionString );
        return new CosmosContainerFactory(client, "Claims");
    })
}

**Comment by gasparnagy** [comment]
@mbhoek have you seen such error before?

**Comment by mbhoek** [comment]
@gasparnagy I haven't, but I'll take a look right away.

**Comment by ajeckmans** [comment]
I actually encountered this in a project as well. I wrote it up to something we did wrong and since we were moving to using solely Bodi I ignored the issue.

**Comment by sanjeev-chevireddy** [comment]
> @gasparnagy It's a duplicate of [solidtoken/SpecFlow.DependencyInjection#88](https://github.com/solidtoken/SpecFlow.DependencyInjection/issues/88)
> 
> @sanjeev-chevireddy Can I ask you, from which version of the SpecFlow plugin are you migrating from?

@mbhoek  3.9.74

**Comment by mbhoek** [comment]
Hmm, that surprises me because I've only ever seen this problem in v3.9.3 of the SpecFlow plugin. I'll investigate.

**Comment by mcraa** [comment]
I encountered the same error earlier but had to make it work quickly so I changed back to the default container.
Later I wanted to reproduce to open an issue, started with a new project and it was working fine. 🤷 

That project was based on the [playwright example from the docs](https://github.com/ZsoltDunai/ReqnrollTestProjectWithPlayWright). Of course, replacing Autofac with MS DI, and using newer versions of most packages.

I will try to switch in that project again to see if I can get back the error.

**Comment by mbhoek** [comment]
@mcraa Thanks that helps a lot! I'm trying to reproduce it in a unit test before solving it.

**Comment by rban-newday** [comment]
Had the same issue when trying to inject `ScenarioContext` in a custom HttpMessageHandler.

**Comment by gasparnagy** [comment]
Another repro for this at: https://github.com/reqnroll/Reqnroll/issues/247#issuecomment-2460231716

**Comment by gasparnagy** [comment]
I have analyzed the issue and found that this error is thrown when someone tries to access a test execution dependent service (e.g. `IReqnrollOutputHelper`) before the tests have started, e.g. in `[BeforeTestRun]` hook. 

I have provided a fix to show a better error message, but in many cases the core problem is an incompatibility with SpecFlow that allowed to access these services in `[BeforeTestRun]` due to a bug (and it was only working for non-parallel execution). Anyway, that problem is discussed in #247 where there is also a workaround provided. The workaround needs to be tweaked though in some cases, see #317).

So this issue is closed now as we have a better error message.
